proposals/0000-hasfield-redesign.rst

```diff
@@ -0,0 +1,1266 @@
+In principle, it is not necessary for ``setField`` or ``modifyField`` to emit a
+runtime error if used with a field that is not present in the datatype; they
+could silently return the value unchanged instead.  This behaviour may be more
+convenient in some circumstances, but may also mask errors, and would not be
+consistent with traditional record updates.
+
+We could imagine giving the option to the user, e.g. via some modifier on the
+datatype definition.  Somewhat related is `proposal #535
+<https://github.com/ghc-proposals/ghc-proposals/pull/535>`_, which suggests an
+extension ``MaybeFieldSelectors`` to control whether partial fields can lead to
+runtime exceptions.
+
+
+Refrain from solving partial fields?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Another option would be for GHC to refrain from solving ``HasField`` or
+``SetField`` constraints automatically where the fields involved are partial.
+This would allow users to define virtual fields with the behaviour they want,
+without conflicting with the automatic solutions.  See `this comment from
+@pnotequalnp <https://github.com/ghc-proposals/ghc-proposals/pull/583#issuecomment-1489278894>`_
+for more motivation for this idea.
+
+However, this would make ``getField`` and ``setField`` less consistent with
+traditional record selectors and record updates.  Moreover it would lead to
+backwards incompatibility for ``HasField``.
+
+
+Affine traversals
+^^^^^^^^^^^^^^^^^
+Optics libraries in principle have a better story to tell here. Partial fields
+give rise to *affine traversals*, where the accessor function returns a
+``Maybe`` value and the setter leaves the value unchanged if it does not mention
+the field (rather than throwing a runtime exception).
+
+We could consider supporting this using built-in classes like the following::
+
+  class GetPartialField x r a | x r -> a where
+    getPartialField :: r -> Maybe a
+
+  class SetPartialField x r a | x r -> a where
+    modifyPartialField :: (a -> a) -> r -> r
+
+  class FieldTotal x r a (is_total :: Bool) | x r -> a is_total
+
+Note that ``modifyField`` will throw an exception on missing fields, whereas
+``modifyPartialField`` would return the value unchanged.  The ``FieldTotal``
+class would allow an optics library to determine whether a particular field was
+total and hence whether it should produce a lens or an affine traversal.
+
+For now we propose not to include support for partial fields through classes
+like this, in the interests of minimizing complexity, and because it is not
+clear how they could be used together with ``OverloadedRecordDot``.
+
+
+Setting vs modification
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A previous iteration of the design supported only ``setField :: a -> r -> r`` and not
+``modifyField :: (a -> a) -> r -> r``.  The latter generalises ``setField`` to
+allow modifying any ``a`` values in the datatype (of which there may be none).
+
+It is easy to implement ``setField`` in terms of ``modifyField``, but not vice
```

Comment:
Is that true even when thinking about memory retention? `setField` can probably drop the reference to the value where `modifyField` can't easily. Do we need `modiftField'`?

Reply:
And performance-worried users would be encouraged to define `setField` explicitly, even if they also define `modifyField`? (Unless the selector thunk optimization applies reliably. Here the think would be `const x y`, right?)

Reply:
In the interest of keeping concerns separate, why not provide `ModifyField` as a separate class entirely?

```haskell
class ModifyField ... where
    modifyField :: (a -> b) -> s -> t
    default modifyField :: (HasField sym s a, SetField sym s b, s ~ t, a ~ b) => (a -> b) -> s -> t
```

This allows potential virtual field use of `SetField` that don't have a corresponding `HasField`, and allows `modifyField` to be optimized in the case where there's a potentially more efficienct solution

Reply:
> And performance-worried users would be encouraged to define `setField` explicitly, even if they also define `modifyField`? (Unless the selector thunk optimization applies reliably. Here the think would be `const x y`, right?)

Thinking about it further, I suppose the selector thunk optimization could help if you have `modifyField` defined in terms of `setField` (because the `getField` selector should be shortcutted) but not vice versa (because I don't think `const` is subject to the optimization, although I wonder if it could be).

I would encourage performance-worried users to make their datatype fields strict. :-)

> why not provide `ModifyField` as a separate class entirely?

We could, but that feels excessive to me. We don't need `modifyField` for `OverloadedRecordUpdate`, so GHC itself wouldn't make use of it. Users defining virtual fields would need to remember to add an extra instance per field (albeit one with a default method) that would only be used in certain situations, and we need laws relating three independent classes whose instances may or may not exist.

I suppose one potentially more compelling option would be to have `ModifyField` support type-changing update (as you hint in your code snippet), and add it independently of `SetField` which could remain non-type-changing. If we wanted to keep `OverloadedRecordUpdate` non-type-changing but add type-changing update support for the optics aficionados, that might not be a terrible approach?